test/test_irods_http_api.py

```diff
@@ -218,6 +218,34 @@ def test_common_operations(self):
         stat_info = r.json()
         self.assertEqual(stat_info['irods_response']['status_code'], -170000) # NOT_A_COLLECTION
 
+    def test_creating_a_collection_with_insufficient_permissions_results_in_an_error_or_no_collection_being_created(self):
+        rodsuser_headers = {'Authorization': 'Bearer ' + self.rodsuser_bearer_token}
+        collection = f'/{self.zone_name}/home/{self.rodsadmin_username}/not_allowed'
+
+        # Attempting to create a collection with insufficient permissions and the
+        # "create-intermediates" parameter not set to 1 results in an error.
+        r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+            'op': 'create',
+            'lpath': collection
+        })
+        #print(r.content) # Debug
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.json()['irods_response']['status_code'], -358000) # OBJ_PATH_DOES_NOT_EXIST
+
+        # Attempting to create a collection with insufficient permissions and the
+        # "create-intermediates" parameter set to 1 results in the target collection
+        # not being created.
+        r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+            'op': 'create',
+            'lpath': collection,
+            'create-intermediates': 1
+        })
+        #print(r.content) # Debug
+        self.assertEqual(r.status_code, 200)
+        result = r.json()
+        self.assertEqual(result['irods_response']['status_code'], 0)
+        self.assertFalse(result['created'])
```

Comment:
When `create-intermediates` is 1, the HTTP API will use `fs::create_collections()`.

This filesystem function does not throw or return errors for many cases. Instead, it returns a boolean indicating whether the collection was created or not.

If the user does not have permission to create the collection, the filesystem function will not report it as an error. It will just say "Hey dude, I didn't create the collection.". Keep in mind, all of the input args could be perfectly valid.

So, for this `create-intermediates=1` case, the status code will always be 0 unless an exception is thrown (which should almost never happen). Users have to inspect the `created` property to determine if something changed.

Does that answer your question?

Reply:
I see, so the fact that created is False is enough - it doesn't give a reason for not doing what was asked of it...

Reply:
feels like... we asked to do a thing, and it didn't... so should be an error returned.   might have to make a table!

Reply:
Can you elaborate? Is there an example use-case which you feel the implemented behavior isn't sufficient?

Reply:
Well, 0 only means no exception was thrown.

When `create-intermediates` is 1, users must check the `created` property to determine whether a collection was created or not.

Alternatively, we put the throwing version in a loop or we design a version for the HTTP API directly.

Reply:
Gut reaction - it feels weird to explain that someone has to check a value other than the return code to see if it 'worked'.  Maybe that's a separate issue to be written up / considered.

Reply:
I think the filesystem lib needs a tweak in that dept. I may have misunderstood what the standard was going for for that library function.

If my understanding is incorrect, then it's easy to extend the implementation to check for specific errors (i.e. permissions) and throw an exception.

I'm pretty sure cppreference made we question the fsys impl which I was looking at this the other day.

Reply:
i know that `mkdir -p` doesn't convey any difference to the user...  but this isn't `imkdir` we're talking about directly, rather the thing that `imkdir` would call...

Reply:
Agreed. The good thing is we can copy the implementation from the fsys lib and build something that fits this project.

It's just a matter of what we want.

Reply:
Updated implementation to use custom version of `fs::create_collections`. Errors will now be reported to the client.

Reply:
oh nice.  resolving for now - will eyeball again on another pass.